test_plans/SYCL-2020/aspect_selector.md

```diff
@@ -0,0 +1,127 @@
+# Test plan for the aspect selector
+
+This is the test plan for the `aspect_selector` feature of SYCL 2020.
+
+## Testing scope
+
+### Device coverage
+
+All tests construct a test device (i.e. `testDevice`),
+representing the device where conformance will be passed on.
+Each test will construct an aspect selector object,
+which will try to select another device,
+this one based on the requested aspects (i.e. `selectedDevice`).
+`selectedDevice` can be any device in the system,
+not necessarily the same as `testDevice`.
+
+### Function overloads
+
+Each test that tests aspects `{asp1, ..., aspN}` must test all three overloads of the `aspect_selector` function:
+
+* Pass a vector of aspects: `aspect_selector(std::vector{asp1, ..., aspN})`
+* Pass aspects as function arguments: `aspect_selector(asp1, ..., aspN)`
+* Pass aspects as template parameters: `aspect_selector<asp1, ..., aspN>()`
+
+### Combinations of aspects
+
+All aspects listed in the `enum class aspect` will be tested,
+let's identify these aspects as `{asp1, ..., aspN}`.
+
+The CTS will cover these combinations:
+
+1. All combinations where no aspect is passed in.
+2. All combinations where a single aspect is used.
+3. All combinations of two aspects,
+including duplicates (e.g. `aspect_selector(aspX, aspX)`),
+and reversed order (e.g. `aspect_selector(aspX, aspY)` and `aspect_selector(aspY, aspX)`).
+
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`,
+where the aspects are not duplicated,
+i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
+The random combinations are selected at test generation time,
+with each test generation producing the same combinations
+and each run testing the same random combinations.
+
+For each of the combinations described so far,
+add new combinations which also use forbidden aspects `{noasp1, ..., noaspM}` -
+`M` is independent of `N`, can be smaller, same size, or larger.
+These forbidden aspects will be passed to the `denyList` argument.
+Use the same rules as above for generating aspects,
+following these extra restrictions:
+
+* `noaspY` is not a member of the set `{asp1, ..., aspN}`
+    for each `noaspY` from the set of forbidden aspects.
```

Comment:
I think this is your intent, but you should make it clear that we will also test the case where the `denyList` is empty.